providers/core.tcpsocket.js

```diff
@@ -296,15 +297,29 @@ Socket_chrome.prototype.dispatchDisconnect = function (code) {
 };
 
 /**
+ * Generates an index key for {@Socket_chrome.active}.
+ * The key is an integer equal to the socket ID for client sockets or, for
+ * server sockets, to the negation of the socket ID.
+ * @method keyForActive
+ * @private
+ * @param {Object} namespace Socket namespace, indicating client or server socket.
+ * @param {number} id The socketId, as provided by chrome.sockets.
+ * @static
+ * @see Socket_chrome.active
+ */
+Socket_chrome.keyForActive = function(namespace, id) {
+  return id * (namespace === chrome.sockets.tcp ? 1 : -1);
```

Comment:
Alternatively structure it as two dictionaries if performance is a worry.
I'm mostly worried about seeing negative ids and confusing them with error
codes or so-forth.

On Tue Jan 13 2015 at 1:33:21 PM trevj <notifications@github.com> wrote:

> In providers/core.tcpsocket.js
> <https://github.com/freedomjs/freedom-for-chrome/pull/56#discussion-diff-22896690>
> :
>
> > @@ -296,15 +297,29 @@ Socket_chrome.prototype.dispatchDisconnect = function (code) {
> >  };
> >
> >  /**
> > + * Generates an index key for {@Socket_chrome.active}.
> > + * The key is an integer equal to the socket ID for client sockets or, for
> > + * server sockets, to the negation of the socket ID.
> > + * @method keyForActive
> > + * @private
> > + * @param {Object} namespace Socket namespace, indicating client or server socket.
> > + * @param {number} id The socketId, as provided by chrome.sockets.
> > + * @static
> > + * @see Socket_chrome.active
> > + */
> > +Socket_chrome.keyForActive = function(namespace, id) {
> > +  return id * (namespace === chrome.sockets.tcp ? 1 : -1);
>
> I certainly can although I figured sticking with integers was prudent for
> performance reasons, given that the key is constructed for every single
> packet received on the connection.
>
> Happy to leave out the (possibly premature) optimisation until we know
> it's a problem, of course. What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/freedomjs/freedom-for-chrome/pull/56/files#r22896690>.
>